Metrics redesign

djdameln opened the pull request:
## 📝 Description

- Adds `Evaluator` class which replaces `MetricsCallback`.
- Similar to `PostProcessor`, the evaluator inherits from `Callback` to allow accessing the val and test outputs, and from `nn.Module` to store the metrics as a submodule in the Lightning model.
- Adds `AnomalibMetric` class, which adds `fields` arg and attribute to existing metrics from torchmetrics. By inheriting from `AnomalibMetric` and a given `Metric` subclass, a new Anomalib-compatible version of the metric is created. When instantiating the metric, the user must specify the batch fields that will be used to update the metric. During update, we only need to pass the batch to the update method.

Some examples:
```python
from anomalib.models import Padim
from anomalib.metrics import AnomalibMetric, AUROC, F1Score
from anomalib.metrics.base import create_anomalib_metric

# A model-specific default set of metrics is provided by anomalib:
# >>> Padim.default_evaluator()
# Evaluator(
#   (val_metrics): ModuleList()
#   (test_metrics): ModuleList(
#     (0): AUROC()
#     (1): F1Score()
#     (2): AUROC()
#     (3): F1Score()
#   )
# )

# But we can also pass our own set of metrics:
image_f1_score = F1Score(fields=["pred_label", "gt_label"])
image_auroc = AUROC(fields=["pred_label", "gt_label"])
Padim(metrics=[image_f1_score, image_auroc])

# When passing multiple metrics of the same type, we need to provide a prefix:
image_f1_score = F1Score(fields=["pred_label", "gt_label"], prefix="image_")
pixel_f1_score = F1Score(fields=["pred_mask", "gt_mask"], prefix="pixel_")
Padim(metrics=[image_f1_score, pixel_f1_score])

# We can also use torchmetrics classes that are not available in Anomalib:
# Option 1: create a custom metric class
class AnomalibAccuracy(AnomalibMetric, Accuracy):
    pass
image_accuracy = AnomalibAccuracy(fields=["pred_label", "gt_label"])

# Option 2: use create_anomalib_metric util function
from torchmetrics import Accuracy
AnomalibAccuracy = create_anomalib_metric(Accuracy)
image_accuracy = AnomalibAccuracy(fields=["pred_label", "gt_label"])

```

Other changes:
- pass `**kwargs` from lightning models to `AnomalyModule` base class

## ✨ Changes

Select what type of change your PR is:

- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🔨 Refactor (non-breaking change which refactors the code base)
- [x] 🚀 New feature (non-breaking change which adds functionality)
- [x] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📚 Documentation update
- [ ] 🔒 Security update

## ✅ Checklist

Before you submit your pull request, please make sure you have completed the following steps:

- [x] 📋 I have summarized my changes in the [CHANGELOG](https://github.com/openvinotoolkit/anomalib/blob/main/CHANGELOG.md) and followed the guidelines for my type of change (skip for minor changes, documentation updates, and test enhancements).
- [ ] 📚 I have made the necessary updates to the documentation (if applicable).
- [ ] 🧪 I have written tests that support my changes and prove that my fix is effective or my feature works (if applicable).

For more information about code review checklists, see the [Code Review Checklist](https://github.com/openvinotoolkit/anomalib/blob/main/docs/source/markdown/guides/developer/code_review_checklist.md).

samet-akcay commented:
@abc-125 @alexriedel1 @blaz-r @jpcbertoldo, if/when you have some time, would you like to review this PR? It introduces some major changes. It would be great to get your perspective.

ashwinvaidya17 commented on src/anomalib/models/components/base/anomaly_module.py:
```diff
@@ -48,10 +52,8 @@ def __init__(self, post_processor: PostProcessor | None = None) -> None:
-        self.pixel_metrics: AnomalibMetricCollection
-
         self.post_processor = post_processor or self.default_post_processor()
+        self.evaluator = self.configure_metrics(metrics)
```
How about we have `self.evaluator = self._setup_evaluator(metrics)`
```python
def _setup_evaluator(metrics) -> Evaulator
```
and
`default_metrics->configure_metrics`?

```python
def configure_metrics() -> Sequence[Metrics] | Metrics:
```

ashwinvaidya17 reviewed (commented):
Except for a discussion regarding naming, I am fine with these changes.
One thing though, how does it look from the CLI, and the config file?
Also, can we use `kwargs` with JSONargparse?

samet-akcay commented on src/anomalib/metrics/f1_score.py:
```diff
@@ -1,36 +1,107 @@
+    """Wrapper to add AnomalibMetric functionality to F1Score metric."""
+
+
+class _F1Max(Metric):
```
I agree, if they are stored in a single module, it should be named something like `f1.py`?

djdameln replied:
I combined these two metrics in a single module because F1Max is just a special case of the F1Score. Both are implementations of the F1 score. One is computed directly over binary predictions while the other one computes the maximum value that can be achieved using the raw anomaly score predictions.

djdameln replied:
The `prefix` argument is useful in any situation where the user wants to compute the same metric multiple times (in different configurations) in a given test run. The most obvious use case is computing the same metric for image and pixel-level predictions, but this is not the only one.

For example, a user could be interested in computing the F1Score for the same set of predictions at different threshold values, e.g.

```python
f1_40 = F1Score(fields=["pred_score", "gt_label"], threshold=0.4, prefix="0.4_")
f1_50 = F1Score(fields=["pred_score", "gt_label"], threshold=0.5, prefix="0.5_")
f1_60 = F1Score(fields=["pred_score", "gt_label"], threshold=0.6, prefix="0.6_")
SomeModel(metrics=[image_f1_score, pixel_f1_score])
```

This is just a quick example of course, but many other use-cases could exist where the same metric type needs to be computed in different configurations.

blaz-r commented:
I'll check this tomorrow in the morning.

blaz-r commented on src/anomalib/metrics/base.py:
```diff
@@ -0,0 +1,117 @@
+
+    def __init__(self, fields: Sequence[str], prefix: str = "", **kwargs) -> None:
+        self.fields = fields
+        self.name = prefix + self.__class__.__name__
```
Maybe box metrics?

djdameln replied:
The `prefix` argument is useful in any situation where the user wants to compute the same metric multiple times (in different configurations) in a given test run. The most obvious use case is computing the same metric for image and pixel-level predictions, but this is not the only one.

For example, a user could be interested in computing the F1Score for the same set of predictions at different threshold values, e.g.

```python
f1_40 = F1Score(fields=["pred_score", "gt_label"], threshold=0.4, prefix="0.4_")
f1_50 = F1Score(fields=["pred_score", "gt_label"], threshold=0.5, prefix="0.5_")
f1_60 = F1Score(fields=["pred_score", "gt_label"], threshold=0.6, prefix="0.6_")
SomeModel(metrics=[image_f1_score, pixel_f1_score])
```

This is just a quick example of course, but many other use-cases could exist where the same metric type needs to be computed in different configurations.

blaz-r reviewed (commented):
I think this looks great 😄. I added some small comments.

I do have two main things to point out:
1. Similar as the one from Ashwin - is there no way to pass the metrics just by name now (CLI, config)?
2. About the default evaluator inside anomaly_module. This change might not be completely compatible with current configs due to early stopping and behavior of existing validation (more details about this in the code comment).

samet-akcay commented:
> I think this looks great 😄. I added some small comments.
> 
> I do have two main things to point out:
> 
> 1. Similar as the one from Ashwin - is there no way to pass the metrics just by name now (CLI, config)?
> 2. About the default evaluator inside anomaly_module. This change might not be completely compatible with current configs due to early stopping and behavior of existing validation (more details about this in the code comment).

@blaz-r, thanks for your review. This PR is part of a greater initiative, which might potentially break things again. That's why we started to think of this as anomalib v2. Here is a proposal showing how Anomalib base model would look like 

[Anomalib - Aux Operations Design.pptx](https://github.com/user-attachments/files/17239491/Anomalib.-.Aux.Operations.Design.pptx)

blaz-r commented:
Cool, thanks for the shared info.

jpcbertoldo reviewed (commented):
IMO `AnomalyModule` should get just
```python
metrics: Evaluator | None
```
instead of
```python
metrics: AnomalibMetric | Sequence[AnomalibMetric] | Evaluator | None
```

From a new user perspective it makes the API a bit easier to understand.

For instance, when `AnomalibMetric | Sequence[AnomalibMetric]` you assume they are test metrics, which is fine, but then you need to warn the user somehow (?)

If the user uses `AnomalibMetric | Sequence[AnomalibMetric]` they already need to be aware of what an `AnomalibMetric` is. Since the init of `Evaluator` is quite simple, you wouldn't make it much harder and at the same time force them to be aware if they're setting val or test.

jpcbertoldo commented:
Just on a side note, if you add `AUPIMO` to this, i would suggest a small change in the design.

Remove the option ` return_average` that we added in the last minute, and instead create another classe `AverageAUPIMO(AUPIMO)`.

It would have something like

```python
def compute(...):
     _, aupimo_result = super().compute(...)
     # normal images have NaN AUPIMO scores
     is_nan = torch.isnan(aupimo_result.aupimos)
     return aupimo_result.aupimos[~is_nan].mean()
```

which is currently done in an `if` inside `AUPIMO`.

djdameln commented on src/anomalib/models/components/base/anomaly_module.py:
```diff
@@ -213,6 +192,42 @@ def default_post_processor(self) -> PostProcessor:
         raise NotImplementedError(msg)
 
+    @staticmethod
+    def default_evaluator() -> Evaluator:
```
I'm personally not a big fan of the `configure_` naming, because it's not an accurate description of how these methods are used. In essence, the `default_evaluator` method (just like the `default_post_processor` method) specifies the default instance of the evaluator that will be used if the user does not pass an evaluator when constructing the model. This is different to `configure_metrics` and `configure_transforms`, which actually contain some configuration logic.

I do agree that there is some inconsistency at the moment between `configure_metrics`, `default_post_processor` and `default_evaluator`, particularly when looking at the `self.evaluator = self.configure_metrics(metrics)` line, which might lead to some confusion. I'll see if I can come up with a more straightforward naming approach, but this might require some discussion.

djdameln commented on src/anomalib/models/components/base/anomaly_module.py:
```diff
@@ -141,29 +143,6 @@ def trainer_arguments(self) -> dict[str, Any]:
         """Arguments used to override the trainer parameters so as to train the model correctly."""
         raise NotImplementedError
 
-    def _save_to_state_dict(self, destination: OrderedDict, prefix: str, keep_vars: bool) -> None:
-        if hasattr(self, "image_threshold"):
-            destination["image_threshold_class"] = (
-                f"{self.image_threshold.__class__.__module__}.{self.image_threshold.__class__.__name__}"
-            )
-        if hasattr(self, "pixel_threshold"):
-            destination["pixel_threshold_class"] = (
-                f"{self.pixel_threshold.__class__.__module__}.{self.pixel_threshold.__class__.__name__}"
-            )
-        if hasattr(self, "normalization_metrics"):
-            for metric in self.normalization_metrics:
-                metric_class = self.normalization_metrics[metric].__class__
-                destination[f"{metric}_normalization_class"] = f"{metric_class.__module__}.{metric_class.__name__}"
-
-        return super()._save_to_state_dict(destination, prefix, keep_vars)
-
-    @staticmethod
-    def _get_instance(state_dict: OrderedDict[str, Any], dict_key: str) -> Threshold:
-        """Get the threshold class from the ``state_dict``."""
-        class_path = state_dict.pop(dict_key)
-        module = importlib.import_module(".".join(class_path.split(".")[:-1]))
-        return getattr(module, class_path.split(".")[-1])()
-
```
Not directly related to the metric refactor, just some cleanup as a part of the overall redesign. Should have probably done this in a separate PR :)

djdameln commented on src/anomalib/models/components/base/export_mixin.py:
```diff
@@ -406,56 +398,12 @@ def val_fn(nncf_model: "CompiledModel", validation_data: Iterable) -> float:
 
         return nncf.quantize_with_accuracy_control(model, calibration_dataset, validation_dataset, val_fn)
 
-    def _get_metadata(
-        self,
-        task: TaskType | None = None,
-    ) -> dict[str, Any]:
-        """Get metadata for the exported model.
-
-        Args:
-            task (TaskType | None): Task type.
-                Defaults to None.
-
-        Returns:
-            dict[str, Any]: Metadata for the exported model.
-        """
-        model_metadata = {}
-        cached_metadata: dict[str, Number | torch.Tensor] = {}
-        for threshold_name in ("image_threshold", "pixel_threshold"):
-            if hasattr(self, threshold_name):
-                cached_metadata[threshold_name] = getattr(self, threshold_name).cpu().value.item()
-        if hasattr(self, "normalization_metrics") and self.normalization_metrics.state_dict() is not None:
-            for key, value in self.normalization_metrics.state_dict().items():
-                cached_metadata[key] = value.cpu()
-        # Remove undefined values by copying in a new dict
-        model_metadata = {key: val for key, val in cached_metadata.items() if not np.isinf(val).all()}
-        del cached_metadata
-        metadata = {"task": task, **model_metadata}
-
-        # Convert torch tensors to python lists or values for json serialization.
-        for key, value in metadata.items():
-            if isinstance(value, torch.Tensor):
-                metadata[key] = value.numpy().tolist()
-
-        return metadata
-
```
yes, same as above. Not related to metrics so you can safely ignore these changes

djdameln commented on src/anomalib/models/components/base/anomaly_module.py:
```diff
@@ -213,6 +192,42 @@ def default_post_processor(self) -> PostProcessor:
+
+        Override in subclass for model-specific evaluator behaviour.
+        """
+        image_auroc = AUROC(fields=["pred_score", "gt_label"], prefix="image_")
+        image_f1score = F1Score(fields=["pred_label", "gt_label"], prefix="image_")
+        pixel_auroc = AUROC(fields=["anomaly_map", "gt_mask"], prefix="pixel_")
+        pixel_f1score = F1Score(fields=["pred_mask", "gt_mask"], prefix="pixel_")
+        test_metrics = [image_auroc, image_f1score, pixel_auroc, pixel_f1score]
+        return Evaluator(test_metrics=test_metrics)
```
In many cases it is not necessary to compute validation metrics. The main purpose of validation in anomaly models is to compute the post-processing metrics such as adaptive threshold and normalization stats, and early stopping in multi-epoch models. The post-processing metrics are covered by the `PostProcessor` class, so the most common reason to compute metrics over the validation set is early stopping. But not all models in Anomalib use or can use early stopping. For example, we have many feature extraction based models which only run for 1 epoch. Computing validation metrics for these models would not be very useful, while it may significantly slow down the training process, as pixel metrics may require some heavy computation.

I don't think the early stopping requirement of some models warrants the unnecessary computation of validation metrics for all other models. If a model uses early stopping by default, then the model implementer should just make sure that the early stopping metric is included by default. If a user wants to enable early stopping for some model, they should add the early stopping metric to the evaluator.

blaz-r replied:
Okay, makes sense. Early stopping is usually not used by original authors, and I agree that this can be handled by the user.

djdameln commented on src/anomalib/models/image/cfa/torch_model.py:
```diff
@@ -227,10 +227,8 @@ def forward(self, input_tensor: torch.Tensor) -> torch.Tensor:
         return InferenceBatch(
             anomaly_map=anomaly_map,
-            pred_score=pred_score,
         )
```
We don't need to compute it here, because we automatically infer the value of `pred_score` from `anomaly_map` for all batches elsewhere.

djdameln replied:
Actually, we might change this as part of the new design, so I'll revert this change. It's out of the scope of this PR anyway.

djdameln commented on src/anomalib/models/components/base/anomaly_module.py:
```diff
@@ -39,7 +39,11 @@ class AnomalyModule(ExportMixin, pl.LightningModule, ABC):
+    def __init__(
+        self,
+        post_processor: PostProcessor | None = None,
+        metrics: AnomalibMetric | Sequence[AnomalibMetric] | Evaluator | None = None,
```
This was my initial design, but we decided to add the possibility of passing metrics directly for a bit more flexibility. 

> If the user uses AnomalibMetric | Sequence[AnomalibMetric] they already need to be aware of what an AnomalibMetric is. Since the init of Evaluator is quite simple, you wouldn't make it much harder and at the same time force them to be aware if they're setting val or test.

One drawback would be that the user needs to import and instantiate `Evaluator`, which would require a bit more knowledge of the code base than just being aware of `AnomalibMetric`. It also increases verbosity, as

```python
image_f1_score = F1Score(fields=["pred_label", "gt_label"])
image_auroc = AUROC(fields=["pred_label", "gt_label"])
Padim(metrics=[image_f1_score, image_auroc])
```

would become
```python
image_f1_score = F1Score(fields=["pred_label", "gt_label"])
image_auroc = AUROC(fields=["pred_label", "gt_label"])
evaluator = Evaluator(test_metrics=[image_f1_score, image_auroc])
Padim(evaluator=evaluator)
```

I do agree that it would simplify the signature of the constructor, and make the class definition more readable. It would also be more in line with the other submodules such as `PostProcessor`. 

@ashwinvaidya17 @samet-akcay any thoughts?

samet-akcay replied:
Looking at the signature again... in terms of naming, which one would make more sense `Metrics` or `Evaluator`,

jpcbertoldo replied:
strongly in favor of ```evaluator: Evaluator | None```

> One drawback would be that the user needs to import and instantiate `Evaluator`, which would require a bit more knowledge of the code base than just being aware of `AnomalibMetric`.

true, but i think it's fine as long as there is tutorial showing the minimal code and a few options (people will just cc cv from there)

samet-akcay commented on tests/unit/utils/test_visualizer.py:
```diff
@@ -48,7 +48,7 @@ def test_model_visualizer_mode(
         """Test combination of model/visualizer/mode on only 1 epoch as a sanity check before merge."""
         _ckpt_path: Path = ckpt_path("Padim")
-        model = get_model("padim")
+        model = Padim(metrics=[])
```
> When the user does not want to compute any metrics at all, they should pass an empty list of metrics as I do here.

Would it be an idea to follow Lightning's `logger` logic, where a user needs to set it to `False` if they explicitly want to disable it.

djdameln commented:
@samet-akcay @ashwinvaidya17 @blaz-r @jpcbertoldo I updated the PR based on your suggestions.

I simplified the `AnomalyModule` signature and logic by only allowing passing an `Evaluator` instance. To run the model without any evaluation, the user can pass `False` (as suggested by @samet-akcay).

I also moved the validation of the input metrics to the `Evaluator` class as suggested by @blaz-r.

blaz-r reviewed (approved):
It does seem like some tests are failing, but once that is fixed I think it's all good. 
Thanks for the effort 😄

djdameln commented:
> Just on a side note, if you add `AUPIMO` to this, i would suggest a small change in the design.
> 
> Remove the option ` return_average` that we added in the last minute, and instead create another classe `AverageAUPIMO(AUPIMO)`.
> 
> It would have something like
> 
> ```python
> def compute(...):
>      _, aupimo_result = super().compute(...)
>      # normal images have NaN AUPIMO scores
>      is_nan = torch.isnan(aupimo_result.aupimos)
>      return aupimo_result.aupimos[~is_nan].mean()
> ```
> 
> which is currently done in an `if` inside `AUPIMO`.

Can you elaborate why this is needed? With this new design we still have access to any constructor arguments of the original metric, so the following would work:
```python
aupimo = AUPIMO(fields=["anomaly_map", "gt_mask"], return_average=False)
```

ashwinvaidya17 reviewed (approved):
Thanks for the efforts! lgtm

jpcbertoldo commented:
> @jpcbertoldo Can you elaborate why this is needed? With this new design we still have access to any constructor arguments of the original metric, so the following would work:
> 
> ```python
> aupimo = AUPIMO(fields=["anomaly_map", "gt_mask"], return_average=False)
> ```

it's not needed : ) 
but it would be more obvious that one is using the _average_ aupimo
also would make a cleaner api IMO

samet-akcay commented on src/anomalib/metrics/evaluator.py:
```diff
@@ -0,0 +1,131 @@
+        val_metrics: AnomalibMetric | Sequence[AnomalibMetric] | None = None,
+        test_metrics: AnomalibMetric | Sequence[AnomalibMetric] | None = None,
+        compute_on_cpu: bool = True,
+        **kwargs,
```
what would be kwargs?

djdameln replied:
kwargs are not needed here. I'll remove them

samet-akcay commented on src/anomalib/models/components/base/anomaly_module.py:
```diff
@@ -48,11 +51,19 @@ def __init__(self, post_processor: PostProcessor | None = None) -> None:
+        # set the post-processor
         self.post_processor = post_processor or self.default_post_processor()
 
+        # set the evaluator
+        self.evaluator: Evaluator | None
+        if isinstance(evaluator, Evaluator):
+            self.evaluator = evaluator
+        elif isinstance(evaluator, bool):
+            self.evaluator = self.configure_evaluator() if evaluator else None
+        else:
+            msg = f"evaluator must be of type Evaluator or bool, got {type(evaluator)}"
+            raise TypeError(msg)
```
As per @ashwinvaidya17's comment on the pre-processor pr, the same goes for this one as well

samet-akcay commented on src/anomalib/models/image/cfa/lightning_model.py:
```diff
@@ -52,8 +52,9 @@ def __init__(
         num_nearest_neighbors: int = 3,
         num_hard_negative_features: int = 3,
         radius: float = 1e-5,
+        **kwargs,
```
What would be the kwargs? My concern is that this could be confusing to users?

samet-akcay reviewed (commented):
I think it looks good. My main concern is whether fields 
- are mandatory to instantiate a metric class
- We use too much `kwargs`, which I think could be confusing to users. It might be an idea to pass explicit args if they are not many.

djdameln commented:
> I think it looks good. My main concern is whether fields
> 
>     * are mandatory to instantiate a metric class

@samet-akcay 

For most metrics, the fields are mandatory to instantiate the metric class. It's the only way to tell the metric which batch fields it should use to update its internal state. There can be some rare exceptions, such as PIMO, where the metric requires a fixed input type by design. In these cases the metric implementer can define the default fields for the metric as a class attribute, like I did [here](https://github.com/djdameln/anomalib/blob/redesign-metrics/src/anomalib/metrics/pimo/pimo.py#L310) for AUPIMO.

samet-akcay reviewed (approved):
lgtm, thanks for the effort
